overlays/nodejs/s3/s3Mixins.ts

```diff
@@ -0,0 +1,232 @@
+        // needed on process `beforeExit`.
+        let subscriptions = bucketSubscriptionInfos.get(bucket);
+        if (!subscriptions) {
+            subscriptions = [];
+            bucketSubscriptionInfos.set(bucket, subscriptions);
+        }
+
+        subscriptions.push({
+            name: name,
+            events: args.events,
+            filterPrefix: args.filterPrefix,
+            filterSuffix: args.filterSuffix,
+            lambdaFunctionArn: this.func.arn,
+            permission: permission,
+        });
+    }
+}
+
+process.on("beforeExit", () => {
+    const copy = bucketSubscriptionInfos;
+
+    // Since we are generating more work on the event loop, we will cause `beforeExit` to be invoked again.
+    // Make sure to clear out eh pending subscriptions array so that we don't try to apply them again.
+    bucketSubscriptionInfos = new Map();
+
+    for (const [bucket, subscriptions] of copy) {
+        const permissions = subscriptions.map(s => s.permission);
+        const _ = new BucketNotification(subscriptions[0].name, {
+            bucket: bucket.id,
+            lambdaFunctions: subscriptions.map(subscription => ({
+                events: subscription.events,
+                filterPrefix: subscription.filterPrefix,
+                filterSuffix: subscription.filterSuffix,
+                lambdaFunctionArn: subscription.lambdaFunctionArn,
+            })),
+        }, { parent: bucket, dependsOn: permissions });
+    }
+});
+
+// Mixin event handling functionality onto Bucket.
+declare module "./bucket" {
+    interface Bucket {
+        onObjectCreated(
+            name: string, handler: BucketEventHandler,
+            args?: ObjectCreatedSubscriptionArgs, opts?: pulumi.ResourceOptions): BucketEventSubscription;
+
+        onObjectRemoved(
+            name: string, handler: BucketEventHandler,
+            args?: ObjectRemovedSubscriptionArgs, opts?: pulumi.ResourceOptions): BucketEventSubscription;
+
+        /**
+         * Creates a new subscription to the given bucket using the lambda provided, along with
+         * optional options to control the behavior of the subscription.  This function should be
+         * used when full control over the subscription is wanted, and other helpers (like
+         * onObjectCreated/onObjectRemoved) are not sufficient.
+         */
+        onEvent(
+            name: string, handler: BucketEventHandler,
+            args: BucketEventSubscriptionArgs, opts?: pulumi.ResourceOptions): BucketEventSubscription;
+    }
+}
+
+Bucket.prototype.onObjectCreated = function (this: Bucket, name, handler, args, opts) {
+    args = args || {};
+    args.event = args.event || "*";
+
+    const argsCopy = {
+        filterPrefix: args.filterPrefix,
+        filterSuffix: args.filterSuffix,
+        events: ["s3:ObjectCreated:" + args.event],
+    };
+
+    return this.onEvent(name + "-object-created", handler, argsCopy, opts);
```

Comment:
It's not clear we need to append `-object-created`, given that the developer is explicitly naming their registration.

Reply:
indeed.  holdover from how we did things before.  removing now.